frontend/src/config/configure.ts

```diff
@@ -0,0 +1,153 @@
+import path from 'path';
+
+import type { BeamerConfig } from '@/types/config';
+
+import type { ChainMetadata } from './chains/chain';
+import type { DeploymentInfo } from './deployment';
+import { TokenMetadata } from './tokens/token';
+import type { BeamerConfigEnvMapping, Environment } from './types';
+import {
+  getEnvironmentForFolder,
+  readAppConfigFile,
+  readChainMetadataFolder,
+  readDeploymentFolder,
+  readTokenMetadataFolder,
+  writeAppConfigFile,
+} from './utils';
+
+const CHAINS_DIR = path.join(__dirname, 'chains');
+const TOKENS_DIR = path.join(__dirname, 'tokens');
+const DEPLOYMENT_DIR = path.join(__dirname, '../../../deployments');
+const OUTPUT_DIR = path.join(__dirname, '../../public');
+
+export const generateTestToken = (rootDeploymentFolder?: string) => {
+  const TST = new TokenMetadata({
+    symbol: 'TST',
+    decimals: 18,
+    addresses: {},
+  });
+
+  const deploymentInfos = readDeploymentFolder(rootDeploymentFolder ?? DEPLOYMENT_DIR, [
+    'mainnet',
+  ]);
+
+  const chainMintableTokenAddressMapping = {};
+
+  for (const info of deploymentInfos)
+    Object.assign(chainMintableTokenAddressMapping, info.getMintableTokenAddresses());
+
+  TST.addAddresses(chainMintableTokenAddressMapping);
+
+  return TST;
+};
+
+/**
+ * Mixes deployment & metadata info
+ */
+export const generateAppConfig = (
+  deploymentInfo: DeploymentInfo,
+  chainMetas: ChainMetadata[],
+  tokenMetas: TokenMetadata[],
+): BeamerConfig => {
+  const config = { chains: {} };
+
+  const chainWithTokens = chainMetas.map((chainMeta) =>
+    chainMeta.formatUsingTokenMetas(tokenMetas),
+  );
+
+  for (const chainId of deploymentInfo.supportedChains) {
+    const chainMeta = chainWithTokens.find((chain) => chain.identifier == Number(chainId));
+    // Avoid mixing chains that have no metadata
+    if (!chainMeta) continue;
+
+    Object.assign(config.chains, {
+      [chainId]: {
+        ...chainMeta,
+        ...deploymentInfo.formatChainDeploymentInfo(chainId),
+      },
+    });
+  }
+  return config;
+};
+
+class ConfigPropertyValueError extends Error {
+  constructor(filePath: string, property: string) {
+    super(`[${filePath}]: Configuration corrupted. Invalid '${property}' value`);
+    Object.setPrototypeOf(this, ConfigPropertyValueError.prototype);
+  }
+}
+// Check if this verify makes sense
```

Comment:
Not sure if this is really necessary. In what case could something go wrong? Would the actual tests be sufficient to verify that correct configs get produced?

Reply:
I found it necessary since we are handling read/writes of `.json` files which are not typed and don't follow any property/type restrictions.

A couple of reasons why:
- Since we are reading the `deployments/${network}/deployment.json` files we need to somehow make sure that those files aren't corrupted/missconfigured for some reason. 
One possible scenario is, when the deployment scripts output a nullish value for some of the properties (most critical is `L2.${chainId}.${contract}.address`
- Reading and using the metadata files from `config/chains/${chainId}.json` & `config/tokens/${tokenSymbol}.json}` can also lead to a missconfiguration since those files are also `.json` without any restrictions on their properties. The chance that the metadata files get missconfigured is even higher since there is manual labor required to maintain them unlike the back-end deployment ones (hopefully in the future we can switch to a well maintained list or repo).
One of the many possible scenarios is when the `rpcUrl` or `explorerTransactionUrl` hold null value. The config generation script will finish uninterrupted but the deployed app will crash when using those corrupted config files.

I could've used type guards here but that would lead to a very generic error report, so that is why i created this function which will provide details for debugging.

The tests cannot guarantee nor test this specifics since the previously described cases will mostly happen on the file input level. Not sure if we are supposed to test the formats of the `.json` files.

Reply:
Ok, now I see. Definitely, there could  something go wrong, because the files are manually changed. 

But as far as I can see, all of this could also be achieved by validating against a JSON schema with a validator such as ajv. This is much easier for ensuring that all properties are set and the types are correct. See https://github.com/ajv-validator/ajv